### PR TITLE
Async DataFrame v1

### PR DESCRIPTION
**Pull request by darth-cheney**
## What ##
This PR implements a _first pass_ at adding async-enabled methods for DataFrames. There are two simultaneous goals here:
1. Add the ability to fetch/store data elsewhere persistently, whether a server or browser storage or whatever. This by definition has to be asynchronous;
2. Offload complex tasks and memory-hogging onto other threads or services or event loops.
  
By _first pass_ here, we mean that we've tried to preserve backward compatibility with the regular `DataFrame` API as much as possible for now. Methods like `getAt` and `putAt` are still synchronous, but will optionally call their async counterparts under the hood and update the `store` as a kind of cache. We might consider refactoring everything in ap-sheet to be asynchronous down the road, hence the "first pass" description.
  
### Replacing generic `callback` with subscribers ###
Previously, any `DataFrame` instance could have a single `callback` property attached to it, which is a function that would be called any time the data changed in the store. This was used to notify the linked GridSheet that something important changed and that data should be refreshed by looking up specific frames or points in the store.
  
We have replaced this with a more general purpose `Set` of "subscribers".
  
Each subscriber is expected to have a method called `onDataChanged` which takes a point or frame as the first argument, and a boolean indicating if a data frame resize occurred as the second argument. It is up to the subscribers as to how they  handle this information. The regular GridSheet does the same thing it did before, dispatching corresponding events as needed and re-rendering its ViewFrame.
  
### Async vs synchronous methods ###
To preserve backwards compatibility for now, DataFrame still implements `getAt` and `putAt` as synchronous functions. They now take an optional (defaulted to true) final argument that indicates whether or not they should "check the async" -- meaning whether they should call the asynchronous version of themselves without waiting for the response (see below).
  
#### Async Versions of accessors ####
The asynchronous versions are called `asyncPutAt` and `asyncGetAt`. In the base DataFrame implementation, they are a no-op. It is expected that subclasses should override and take some action. The idea is to then notify any subscribers that they need to check the `store`, as data may have been updated there via these asynchronous requests. These can take any form, and the `IDBDataFrame` subclass is just one example.
  
#### Other async call refactorings ####
Other functions that are frequently used have been refactored to be asynchronous *without maintaining a synchronous form*. These include `loadFromArray`, `copy` methods, and several others. See code changes for more information.
  
### `IDBDataFrame` ###
As an example, we have included a subclass of the new async-enabled `DataFrame` called `IDBDataFrame`. The purpose of this class is to override the asynchronous methods in `DataFrame` such that it uses IndexedDB as a persistent data store.
  
In this branch we have updated the plain `GridSheet` to use `IDBDataFrame`.
  
**NOTE**: In a future PR, we should  add some methods for setting different "kinds" of DataFrames on a GridSheet.
  
## Tests ##
Many tests have been updated to use the new asynchronous versions of methods and all are passing.

## Trying out ##
[http://localhost:8000/examples](http://localhost:8000/examples) should be enabled with the example IndexedDB backed IDBDataFrame. You can use the devtools to explore the databse values as they change.

### CODE REVIEW

File: src/GridSheet.js [hunk 1]
```diff
@@ -233,6 +226,9 @@ class GridSheet extends HTMLElement {
         this.addEventListener("selection-changed", this.handleSelectionChanged);
         this.addEventListener("sheet-view-shifted", this.handleViewShift);
         this.addEventListener("cell-edited", this.handleCellEdited);
+
+        // Add the sheet itself as a subscriber to its DataFrame
+        this.dataFrame.subscribers.add(this);
```
Comment (dkrasner) [inline src/GridSheet.js:231]: how come GridSheet doesn't have a `onDataChanged()` ?

**Comment by dkrasner** [comment]
@darth-cheney can you be a bit more specific about how to test this out. There is one example `async-testing-1.html` that i can see, which loads the classes but has an empty document.body

File: src/GridSheet.js (hunk 1 above)
Comment (darth-cheney) [reply src/GridSheet.js:231]: It does. It's [here](https://github.com/darth-cheney/ap-sheet/pull/48/files/e361f15fd7b9bdab2208e920960209d8e08831d2#diff-ad6cea5532b3c5febcaf872e88b3c81b25cf47e0ac8d32b91590eeab616b0cf6R266)

**Comment by darth-cheney** [comment]
@dkrasner if you run a simple http server at the root of the repo, then http://localhost:8000/examples will load an example that is wired for IDBDataFrame (indexeddb). You can use the browser dev tools to see the data update in indexed db also as you make changes (it's the "Application" tab in Chrome, "Storage" tab in Firefox)

**Comment by dkrasner** [comment]
yeah i was looking at that, but not exactly sure what i am testing here...

can you make a sheet example, ie a sheet that works off of idb, so then you can change the sheet values, see how that interacts with the store, db etc

**Comment by dkrasner** [comment]
Ah i see, i kept looking at the async* example. Cool. Seems to be working. It's insane that the debugger is not responsive for IDB, ie I need to click out/in of defaultDB/cells to see the update.. .uggh

clearing is throwing an error:
image


but `dataFrame.clear(true)` does not and clears data from IDB. 

However, it does not clear data from the store. 

Try this:
add some values to the sheet, make sure all is good and you see them in the store and in IDB

then run:
```
temp0.dataFrame.clear(true)
temp0.dataFrame.loadFromArray([[1, 3], [5, 6]])
```

you'll see the cleared values resurrected

**Pull request merged**